hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/replication/LegacyReplicationManager.java

```diff
@@ -523,22 +523,32 @@ protected void processContainer(ContainerInfo container,
         boolean sufficientlyReplicated = replicaSet.isSufficientlyReplicated();
         boolean placementSatisfied = placementStatus.isPolicySatisfied();
         if (!sufficientlyReplicated || !placementSatisfied) {
-          if (!sufficientlyReplicated) {
+          // Increment report stats.
+          if (!sufficientlyReplicated && replicaSet.isUnrecoverable()) {
+            report.incrementAndSample(HealthState.MISSING,
+                container.containerID());
             report.incrementAndSample(
                 HealthState.UNDER_REPLICATED, container.containerID());
-            if (replicaSet.isUnrecoverable()) {
-              report.incrementAndSample(HealthState.MISSING,
-                  container.containerID());
-            }
           }
           if (!placementSatisfied) {
             report.incrementAndSample(HealthState.MIS_REPLICATED,
                 container.containerID());
 
           }
+          // Replicate container if needed.
           if (!inflightReplication.isFull() || !inflightDeletion.isFull()) {
-            handleUnderReplicatedContainer(container,
-                replicaSet, placementStatus);
+            if (!replicaSet.isUnrecoverable()) {
+              if (replicaSet.getHealthyReplicaCount() == 0 &&
+                  replicaSet.getUnhealthyReplicaCount() != 0) {
+                handleAllReplicasUnhealthy(container, replicaSet,
+                    placementStatus, report);
+              } else {
+                report.incrementAndSample(
+                    HealthState.UNDER_REPLICATED, container.containerID());
+                handleUnderReplicatedHealthy(container,
+                    replicaSet, placementStatus);
+              }
+            }
```

Comment:
I think this logic will not update this container's under replicated state in the report if it's recoverable and those maps are full

Reply:
Good catch. I fixed this in [fe2fd7e](https://github.com/apache/ozone/pull/3920/commits/fe2fd7e0b047c6f1b06e9b5b8ecb1f9f5fc07600) by having a dedicated method to generate the replication related parts of the report. Having them mixed with the handlers for these cases was messy.